python/pycrdt/_awareness.py

```diff
@@ -38,164 +36,148 @@ def states(self) -> dict[int, dict[str, Any]]:
-        Apply states update and sends the changes to subscribers.
-
-        Args:
-            message: The binary changes.
-            origin: The origin of the change.
+        return self._states.get(self.client_id)
 
-        Returns:
-            A dictionary with the changes and the origin.
-        """
-        update = read_message(update)
-        decoder = Decoder(update)
-        states = []
-        length = decoder.read_var_uint()
-        states_changes: dict[str, list[int]] = {
-            "added": [],
-            "updated": [],
-            "filtered_updated": [],
-            "removed": [],
-        }
-
-        for _ in range(length):
-            client_id = decoder.read_var_uint()
-            clock = decoder.read_var_uint()
-            state_str = decoder.read_var_string()
-            state = None if not state_str else json.loads(state_str)
-            if state is not None:
-                states.append(state)
-            self._update_states(client_id, clock, state, states_changes)
-
-        changes = {
-            "changes": states_changes,
-            "origin": origin,
-        }
-
-        # Do not trigger the callbacks if it is only a keep alive update
-        if (
-            states_changes["added"]
-            or states_changes["filtered_updated"]
-            or states_changes["removed"]
-        ):
-            for callback in self._subscriptions.values():
-                callback(changes)
-
-        return changes
-
-    def get_local_state(self) -> dict[str, Any]:
-        """
-        Returns:
-            The local state.
-        """
-        return self._states.get(self.client_id, {})
-
-    def set_local_state(self, state: dict[str, Any]) -> dict[str, Any]:
+    def set_local_state(self, state: dict[str, Any] | None) -> None:
         """
         Updates the local state and meta, and sends the changes to subscribers.
 
         Args:
-            state: The new local state.
-
-        Returns:
-            A dictionary with the changes and the origin (="local").
+            state: The new local state, if any.
         """
-        clock = self._meta.get(self.client_id, {}).get("clock", 0) + 1
-        states_changes: dict[str, list[int]] = {
-            "added": [],
-            "updated": [],
-            "filtered_updated": [],
-            "removed": [],
-        }
-        self._update_states(self.client_id, clock, state, states_changes)
-
-        changes = {
-            "changes": states_changes,
-            "origin": "local",
-        }
-
-        if (
-            states_changes["added"]
-            or states_changes["filtered_updated"]
-            or states_changes["removed"]
-        ):
+        client_id = self.client_id
+        curr_local_meta = self._meta.get(client_id)
+        clock = 0 if curr_local_meta is None else curr_local_meta["clock"] + 1
+        prev_state = self._states.get(client_id)
+        if state is None:
+            del self._states[client_id]
```

Comment:
This can probably raise an error if the local state has not been defined and we try to set it as None.

Reply:
Raising an exception is a feature IMO (rather than keep it silent), but maybe you mean that we should raise a more meaningful exception?

Reply:
I don't think we need to raise an exception, if we want to set it to null but it is already the case, it might be silent, like [here](https://github.com/jupyter-server/pycrdt/pull/170/files#diff-d0908cdc750b8113ae6b4433e378db136aebe96a30e773aef2cf08313b6669d8R152)

Reply:
But in `set_local_state` we are responsible for what we do, in `apply_awareness_update` less so. But I don't have a strong opinion on this, I will add the check back :+1: